Match non-linear overscroll spring to iOS

xster opened the pull request:
Tested on all iOS devices (i.e. the resistance increase is indeed related to the portion of the viewport overscrolled)

#10972

**Before**:
![img_0435 1](https://user-images.githubusercontent.com/156888/27758109-1170a1ec-5dc2-11e7-861b-2e3ab3dbb3c2.gif)

**After**:
![img_0431](https://user-images.githubusercontent.com/156888/27758127-5e91d798-5dc2-11e7-896d-b137c5c92372.gif)
![img_0432](https://user-images.githubusercontent.com/156888/27758125-5d2a0d9e-5dc2-11e7-96b2-1307ca47d60e.gif)

xster committed:
Make the drag resistance non-linear

Hixie commented on packages/flutter/lib/src/widgets/scroll_physics.dart:
```diff
@@ -227,17 +227,29 @@ class BouncingScrollPhysics extends ScrollPhysics {
   double applyPhysicsToUserOffset(ScrollMetrics position, double offset) {
     assert(offset != 0.0);
     assert(position.minScrollExtent <= position.maxScrollExtent);
     if (offset > 0.0)
```
nit: braces

xster replied:
Refactored

Hixie commented:
Test?

xster committed:
Let the easing of overscroll have a spring effect too

xster commented:
Just added the fix for the other bug in #10972 in the same PR

Easing off from an overscroll used to just move normally whereas the iOS scroll still applies the spring in reverse during the ease.

Hixie commented on packages/flutter/lib/src/widgets/scroll_physics.dart:
```diff
@@ -227,32 +227,41 @@ class BouncingScrollPhysics extends ScrollPhysics {
-        return total + delta * gamma;
-      total += distanceToLimit;
-      delta -= deltaToLimit;
+    final double distanceOutsideOfLimit = start < lowLimit
```
I have no idea what is going on here.

Hixie replied:
(Which is fine, not asking for a change or anything.)

xster replied:
Split out more variables

eseidelGoogle commented:
I am *soo* excited for this change.  Thank you!

HansMuller commented on packages/flutter/lib/src/widgets/scroll_physics.dart:
```diff
@@ -227,32 +227,41 @@ class BouncingScrollPhysics extends ScrollPhysics {
-  /// scroll.
-  double get frictionFactor => 0.5;
+  /// This factor starts at 0.54 and progressively becomes harder to overscroll
+  /// as more of the area past the edge is dragged in.
```
This seems like the heart of a good change. It might be possible to explain it a little more clearly. Is "inViewFraction" the percentage of scrollable content that's still visible?

xster replied:
Added more doc

HansMuller commented on packages/flutter/lib/src/widgets/scroll_physics.dart:
```diff
@@ -227,32 +227,41 @@ class BouncingScrollPhysics extends ScrollPhysics {
-    if (offset > 0.0)
-      return _applyFriction(position.pixels, position.minScrollExtent, position.maxScrollExtent, offset, frictionFactor);
-    return -_applyFriction(-position.pixels, -position.maxScrollExtent, -position.minScrollExtent, -offset, frictionFactor);
+    final double direction = offset.sign;
```
NICE. A variable introduced to help the reader!

HansMuller replied:
What's going on here isn't obvious to me either. Is "distanceOutsideOfLimit" how far we're overscrolled? It's not your fault that there's a parameter called "gamma" but it would be nice to know what gamma means. A short comment that explains what's happening here would be be helpful.

xster replied:
Unfortunately I not sure which formula was being referred to originally with gamma either :(

Though I refactored a bit with more extracted variables so it's hopefully a bit easier to follow.

HansMuller commented:
LGTM. Make @eseidel  happy!

xster committed:
Add tests and prevent possible drift by having a slightly smaller resistance when easing the overscroll

xster commented:
Added tests and added another fine tuning to prevent a possible drift problem if the user dragged up and down repeatedly while in overscroll the whole time. 

Some readability refactor while doing so.

xster committed:
lint
